### PR TITLE
Different improvements for AMS

### PR DESCRIPTION
**Pull request by peterrum**
- setup: compute weights during initialization and convert indices from global to local
- add timers
- reused smoother preconditioner instances

### CODE REVIEW

**Review by lpsaavedra — APPROVED** [review]

File: source/solvers/fluid_dynamics_matrix_free.cc [hunk 1]
```diff
@@ -170,14 +217,36 @@ class PreconditionASM : public PreconditionBase<VectorType>
...
+        this->patches[c].reserve(patches[c].size());
+
+        for (const auto &i : patches[c])
+          this->patches[c].emplace_back(partition->global_to_local(i));
+      }
 
-    src.reinit(partition);
-    dst.reinit(partition);
+    this->timer.leave_subsection("asm::inidices");
+    this->timer.enter_subsection("asm::restrict");
```
Comment (blaisb) [inline source/solvers/fluid_dynamics_matrix_free.cc:249]:
```suggestion
    this->timer.leave_subsection("asm::indices");
    this->timer.enter_subsection("asm::restrict");
```

File: source/solvers/fluid_dynamics_matrix_free.cc [hunk 2]
```diff
@@ -230,7 +293,28 @@ class PreconditionASM : public PreconditionBase<VectorType>
...
                                                         (1.0 / i);
         weights.update_ghost_values();
+
+        this->timer.leave_subsection("asm::weight");
       }
+  }
+
+  /**
+   * @brief Apply preconditioner.
+   */
```
Comment (blaisb) [inline source/solvers/fluid_dynamics_matrix_free.cc:303]: Can you add @param for the parameters

File: source/solvers/fluid_dynamics_matrix_free.cc [hunk 3]
```diff
@@ -52,11 +53,48 @@ template <typename VectorType>
 class PreconditionBase
 {
 public:
+  /**
+   * @brief Constructor.
+   */
```
Comment (blaisb) [inline source/solvers/fluid_dynamics_matrix_free.cc:58]: can you add @param

File: source/solvers/fluid_dynamics_matrix_free.cc [hunk 4]
```diff
@@ -52,11 +53,48 @@ template <typename VectorType>
...
+  PreconditionBase()
+    : comm(MPI_COMM_WORLD)
+    , pcout(std::cout, Utilities::MPI::this_mpi_process(comm) == 0)
+    , timer(pcout, TimerOutput::never, TimerOutput::wall_times)
+  {}
+
   /**
    * @brief Apply preconditioner.
    */
```
Comment (blaisb) [inline source/solvers/fluid_dynamics_matrix_free.cc:67]: Can you add @param

**Review by blaisb — CHANGES_REQUESTED** [review]
Just a few things to fix

**Comment by peterrum** [comment]
@blaisb Ready from my side!

**Review by blaisb — APPROVED** [review]
Awesome. I rebased. Will merge after CI.

**Pull request merged**